src/main/java/pl/edu/mimuw/BigInt.java

```diff
@@ -21,22 +35,199 @@ public BigInt(int number) {
     this(Integer.toString(number));
   }
 
+  //trim przyjmuje liczbe ktora moze miec zera na poczatku i zwraca taka bez zer
+  public BigInt trim(){
+    int len = this.digits.length;
+    int realLen = len;
+    
+    while((this.digits[realLen-1]==0)&&(realLen>1)){
+      realLen--;
+    }
+    int[] newDigits = new int[realLen];
+    for(int i=0; i<realLen; i++){
+      newDigits[i]=digits[i];
+    }
+    //jesli w wyniku pojawi sie samo "0", to zmieniamy znak na dodatni, by uniknac roznych reprezentacji zera
+    if(realLen==1){
+      if(newDigits[0]==0){
+        return new BigInt(newDigits, true);
+      }
+    }
+    return new BigInt(newDigits, this.isPositive);
+  }
+
   public BigInt neg() {
     final var digits = Arrays.copyOf(this.digits, this.digits.length);
     return new BigInt(digits, !this.isPositive);
   }
 
+
   public BigInt add(BigInt other) {
-    throw new IllegalStateException("TODO task 2: implement addition");
+    boolean newSign=true;
+    int[] newDigits;
+    int carriage=0;
+    int sum = 0;
+    int len=java.lang.Math.max(other.digits.length, this.digits.length);
+
+    //dzielimy na dwa przypadki - tych samych i roznych znakow
+    //te same znaki
+    if(this.isPositive==other.isPositive){
```

Comment:
o jejku :D to powinno być rozbite na conajmniej 2 metody pomocnicz, sensownie nazwane ;) trudno tak sie polapac, mimo komentarzy. Polecam w intellij skrot Ctrl+Alt+M na zaznaczonym kodzie do ekstrakcji metody